parsec/core/fs/workspacefs/entry_transactions.py

```diff
@@ -198,17 +199,30 @@ async def entry_missing_data(
                     total_size += chunk.raw_size
 
         block_ids = [chunk.id for chunk in blocks]
-        missing_size = 0
 
-        local_chunk_ids = await self.local_storage.block_storage.get_local_chunk_ids(block_ids)
+        local_blocks = []
+        remote_blocks = []
+        local_and_remote_blocks = []
+
+        # To avoid concurrency problems local storage is called first
+        local_and_remote_chunk_ids = await self.local_storage.block_storage.get_local_chunk_ids(
+            block_ids
+        )
+        local_and_remote_chunk_ids_set = set(local_and_remote_chunk_ids)
+
+        local_chunk_ids = await self.local_storage.chunk_storage.get_local_chunk_ids(block_ids)
```

Comment:
I would call those variables:
- `block_ids`
- `local_and_remote_block_ids`
- `local_only_block_ids`

And then compute `remote_only_block_ids` as:
```python
remote_only_block_ids = block_ids - local_and_remote_block_ids - local_only_block_ids
```
Also if you use a mapping id-to-block called `block_dict` you can then do:
```python
local_only_blocks = [block_dict[block_id].access for block_id in local_only_block_ids]
remote_only_blocks = [block_dict[block_id].access for block_id in remote_only_block_ids]
local_and_remote_blocks = [block_dict[block_id].access for block_id in local_and_remote_block_ids]
```

To sum it up:
```python
block_dict = {chunk.id: chunk for chunks in manifest.blocks for chunk in chunks if chunk.is_block()}
block_ids = set(block_dict)

local_and_remote_chunk_ids = set(await self.local_storage.block_storage.get_local_chunk_ids(block_ids))
local_only_chunk_ids = set(await self.local_storage.chunk_storage.get_local_chunk_ids(block_ids))
remote_only_block_ids = block_ids - local_and_remote_block_ids - local_only_block_ids

local_only_blocks = [block_dict[block_id].access for block_id in local_only_block_ids]
remote_only_blocks = [block_dict[block_id].access for block_id in remote_only_block_ids]
local_and_remote_blocks = [block_dict[block_id].access for block_id in local_and_remote_block_ids]
```
What do you think?

Reply:
block_dict needs to not be a list comprehension to be able to break when we reach the limit argument passed to the function apart from that LGTM